### ISSUE TITLE
Why request block body for fast sync?

### ISSUE THREAD

**Issue by liuchengxu** [paritytech/polkadot-sdk#5406]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

If I read correctly,  the docs https://github.com/paritytech/polkadot-sdk/blob/3fe22d17c4904c5ae016034b6ec2c335464b4165/substrate/client/network/README.md#fast-sync indicates that the fast sync is expected only to download and verify the header chain, and the subsequent state snapshot.

But the inline docs are contradictory, saying Downloading blocks:
https://github.com/paritytech/polkadot-sdk/blob/3fe22d17c4904c5ae016034b6ec2c335464b4165/substrate/client/cli/src/arg_enums.rs#L275-L278

and the block body is indeed requested for `--sync fast` and `--sync fast-unsafe`.

https://github.com/paritytech/polkadot-sdk/blob/3fe22d17c4904c5ae016034b6ec2c335464b4165/substrate/client/network/sync/src/strategy/chain_sync.rs#L1220-L1225

I'm looking for a sync mode that only downloads the headers and the state snapshot. I initially assumed that the --sync fast/fast-unsafe options would fulfill this requirement. This looks like a bug to me, could you please clarify this confusion?

**Comment by liuchengxu** [comment]
After looking into the original PR of fast sync https://github.com/paritytech/substrate/pull/8884, I found this comment https://github.com/paritytech/substrate/pull/8884#issuecomment-847712287, so it was intentionally to request the body in fast sync. However, this is a mismatch between the implementation and [spec](https://spec.polkadot.network/chap-sync#sect-sync-fast) IMHO. The existing documentation always mentions downloading the header chain, but the block history is downloaded _silently_. While it may not be a big problem for the chain without a huge block history, for a project like [subcoin](https://github.com/subcoin-project/subcoin), it matters about downloading hundreds of MiB or hundreds of GiB data before downloading the state. We should fix this mismatch in the implementation, what do you think? @arkpar

**Comment by arkpar** [comment]
Indeed, fast sync is supposed to download and verify the header chain first, then the state, then block bodies in the background.

**Comment by liuchengxu** [comment]
Great that we're on the same page! Any insights on the implementation? If I read the code correctly, the gap sync will be started automatically to download the block history once the block gap is detected in `client.info().block_gap` within the chain sync restart function. How about reusing the gap sync to download the block bodies by updating the `block_gap` for fast sync? @arkpar

**Comment by arkpar** [comment]
Right, the backgroudn block body download should utilize the same mechanism as the warp sync. The gap in this case is in block bodies, rather than whole block. The `Client` should have no issue importing a body for an existing block header.
@dmitry-markin can give you further hints, as I have not touched that code for a while.

**Comment by dmitry-markin** [comment]
Going through `Backend::try_commit_operation()` it looks like if block header is imported, it does not considered to be a block gap anymore. So gap sync probably won't run after fast sync because no gap will be detected. Makes sense double checking by not requesting block bodies during fast sync and checking if gap sync was performed.

**Comment by liuchengxu** [comment]
So I guess we need to either change the logic of detecting the block gap within `Backend::try_commit_operation()` or update the block gap manually after the state sync is finished (and ensure the Client can import the body with an existing header) so that the gap sync can be started as before. The former seems to be more straightforward.

**Comment by dmitry-markin** [comment]
Yes, the first approach is better. Worth checking where else the block gap is used and if it won't break anything.

**Comment by liuchengxu** [comment]
We need a formal definition of a block gap before implementation. The concept of a block gap has evolved. Initially, it could only be created during warp sync; now, it can be created by both warp sync and fast sync. Warp sync creates a block gap by first downloading the finality proof and state at a specific height and then proceeding to download the block history. In contrast, fast sync downloads the header chain and state at a specific height before downloading the block history.

#### Block Gap Creation and Updates

- Warp Sync
    - Creating a Block Gap
        - A block gap is created when the warp block is imported and the following condition is met: `!parent_header_exists && number > best_num + 1` https://github.com/paritytech/polkadot-sdk/blob/da11c7e9122d1d695a3dea57c34a73815faecf01/substrate/client/db/src/lib.rs#L1710-L1711
    - Updating a Block Gap (During Block History Download)
        - The block gap is decreased when a block in the history is downloaded
            - Condition: `number == block_gap.start` https://github.com/paritytech/polkadot-sdk/blob/da11c7e9122d1d695a3dea57c34a73815faecf01/substrate/client/db/src/lib.rs#L1688
            - Action: `gap.start += 1` => **gap decreased** https://github.com/paritytech/polkadot-sdk/blob/da11c7e9122d1d695a3dea57c34a73815faecf01/substrate/client/db/src/lib.rs#L1689
- Fast Sync
    - Creating a Block Gap
        - A block gap is created when a new header is imported and the following condition is met: `parent_header_exists && !body_exists && number == best_num + 1`
    - Updating a Block Gap (During Header Chain and Block History Download)
        - Header Chain Download: The block gap is increased when a new header is downloaded:
            - Condition: `number == gap.end + 1 && !body_exists`
            - Action: `gap.end += 1` => **gap increased**
        - Block History Download: The block gap is decreased when a block in the history is downloaded:
            - Condition `number == block_gap.start`
            - Action: `gap.start += 1` => **gap decreased**

#### Block Gap Check

- Previous check: `!existing_header`. https://github.com/paritytech/polkadot-sdk/blob/da11c7e9122d1d695a3dea57c34a73815faecf01/substrate/client/db/src/lib.rs#L1658
- New check: `!existing_header || !body_exists`.

The logic for decreasing a block gap is consistent across both warp sync and fast sync, but fast sync can increase the gap during the header chain download.

#### Implementation Plan

1. Refactor `BlockGap`.

The current BlockGap is represented as `Option<(NumberFor<Block>, NumberFor<Block>)>`. To improve extensibility, this tuple should be refactored into a struct. Compatibility concerns should be addressed by upgrading the old gap storage to the new structure.

```Rust
pub enum BlockGapType {
    // Gap created by warp sync.
    MissingHeaderAndBody,
    // Gap created by fast sync.
    MissingBody,
}

pub struct BlockGap<N> {
    start: N,
    end: N,
    gap_type: BlockGapType,
}
```

2. Add Block Gap Support for Fast Sync.

The primary changes will occur in `try_commit_operation`, with adjustments to the ChainSync logic as well.

I've implemented a working prototype locally based on the ideas above, and it appears to be functioning well. I would appreciate it if you could review the changes and provide feedback in case I missed something important. @arkpar @dmitry-markin

**Comment by dmitry-markin** [comment]
@liuchengxu sure, feel free to publish your local changes as a PR / draft PR.

**Comment by liuchengxu** [comment]
Please have a look at https://github.com/paritytech/polkadot-sdk/pull/5540 before I send a new PR @dmitry-markin .
